File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/utils/WeeksTest.scala
package com.github.debop4s.timeperiod.tests.utils

import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.utils.Times._
import com.github.debop4s.timeperiod.utils.{Weeks, Times}
import org.joda.time.Duration

/**
 * com.github.debop4s.timeperiod.tests.utils.WeeksTest
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2014. 2. 17. 오전 11:45
 */
class WeeksTest extends AbstractTimePeriodTest {

  val testTimes = Array(asDate(2003, 12, 28), asDate(2014, 12, 30))

  test("year and week") {
    testTimes.foreach { m =>
      val yw = Weeks.getYearAndWeek(m)
      yw.weekyear should equal(m.getWeekyear)
      yw.weekOfWeekyear should equal(m.getWeekOfWeekyear)
    }
  }

  test("parallel year and week") {
    (2000 to 2100).par.foreach { year =>
      val startDay = Times.startTimeOfYear(year)
      val endDay = Times.endTimeOfYear(year - 1)

      val startYW = Weeks.getYearAndWeek(startDay)
      val endYW = Weeks.getYearAndWeek(endDay)

      if (startDay.getDayOfWeek == FirstDayOfWeek.id)
        endYW should not equal startYW
      else
        endYW should equal(startYW)
    }
  }

  test("start week range of year") {
    (2000 to 2100).par.foreach(year => {
      val startWR = Weeks.getStartWeekRangeOfYear(year)
      log.trace(s"year=$year, startWeek=${startWR.startDayStart }")

      new Duration(asDate(year - 1, 12, 28), startWR.startDayStart).getStandardDays should be > 0L
      new Duration(asDate(year, 1, 3), startWR.endDayStart).getStandardDays should be > 0L
    })
  }

  test("endYearAndWeek") {
    (1980 to 2200).par.foreach { year =>
      val yw = Weeks.getEndYearAndWeek(year)
      yw.weekyear should equal(year)
      yw.weekOfWeekyear should be >= 52
    }
  }

  test("end week range of year") {
    (2000 to 2100).par.foreach { year =>
      val startWR = Weeks.getStartWeekRangeOfYear(year)
      val endWR = Weeks.getEndWeekRangeOfYear(year - 1)
      log.trace(s"year=$year, startWR=${startWR.startDayStart }, endWR=${endWR.startDayStart }")

      new Duration(asDate(year - 1, 12, 28), startWR.startDayStart).getStandardDays should be > 0L
      new Duration(asDate(year, 1, 3), startWR.endDayStart).getStandardDays should be > 0L

      endWR.startDayStart.plusWeeks(1) should equal(startWR.startDayStart)
      endWR.endDayStart.plusDays(1) should equal(startWR.startDayStart)
    }
  }

  test("get week range") {
    (2000 to 2100).par.foreach { year =>
      val endDay = Times.endTimeOfYear(year - 1)
      val startDay = Times.startTimeOfYear(year)

      val endDayYearWeek = Weeks.getYearAndWeek(endDay)
      endDayYearWeek.weekyear should be >= year - 1

      val startDayYearWeek = Weeks.getYearAndWeek(startDay)
      startDayYearWeek.weekyear should be <= year

      // 해당일자가 속한 주차의 날짜들을 구한다. 년말/년초 구간은 꼭 7일이 아닐 수 있다.
      val endDayWeekRange = Weeks.getWeekRange(endDayYearWeek)
      val startDayWeekRange = Weeks.getWeekRange(startDayYearWeek)

      log.trace(s"start day weeks=$startDayWeekRange")

      if (endDayYearWeek == startDayYearWeek)
        startDayWeekRange should equal(endDayWeekRange)
      else
        startDayWeekRange should not equal endDayWeekRange
    }
  }

  test("add WeekOfYears") {
    (2000 to 2100).par.foreach { weekyear =>
      val step = 2
      val maxAddWeeks = 40

      var prevResult: YearWeek = null
      val maxWeek = Weeks.getEndYearAndWeek(weekyear)

      for (week <- 1 until maxWeek.weekOfWeekyear by step) {
        for (addWeeks <- -maxAddWeeks until maxAddWeeks by step) {
          val current = YearWeek(weekyear, week)
          val result = Weeks.addWeekOfYears(current, addWeeks)

          if (addWeeks != 0 && prevResult != null) {
            if (result.weekyear == prevResult.weekyear)
              result.weekOfWeekyear should equal(prevResult.weekOfWeekyear + step)
          }

          result.weekOfWeekyear should be > 0
          result.weekOfWeekyear should be <= MaxWeeksPerYear

          prevResult = result
        }
      }
    }
  }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/tools/WeeksTest.java
package debop4s.timeperiod.tests.tools;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.YearWeek;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.WeekRange;
import debop4s.timeperiod.utils.Times;
import debop4s.timeperiod.utils.Weeks;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.joda.time.Duration;
import org.junit.Test;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.tools.WeeksFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 25. 오후 12:55
 */
@Slf4j
public class WeeksTest extends TimePeriodTestBase {

    public static final DateTime[] testTimes = new DateTime[] {
            Times.asDate(2003, 12, 28)
    };

    @Test
    public void getYearAndWeek() {
        for (DateTime moment : testTimes) {
            YearWeek yw = Weeks.yearAndWeek(moment);
            assertThat(yw.weekyear()).isEqualTo(moment.getWeekyear());
            assertThat(yw.weekOfWeekyear()).isEqualTo(moment.getWeekOfWeekyear());
        }
    }

    @Test
    public void getYearAndWeekTest() {

        JParallels.run(2000, 2100, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {
                DateTime startDay = Times.startTimeOfYear(year);
                DateTime endDay = Times.endTimeOfYear(year - 1);

                YearWeek startYW = Weeks.yearAndWeek(startDay);
                YearWeek endYW = Weeks.yearAndWeek(endDay);

                if (startDay.getDayOfWeek() == TimeSpec.FirstDayOfWeek.getValue())
                    assertThat(endYW.equals(startYW)).isFalse();
                else
                    assertThat(endYW.equals(startYW)).isTrue();
            }
        });
    }

    @Test
    public void getStartWeekRangeOfYear() {
        JParallels.run(2000, 2100, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {
                WeekRange startWeekRange = Weeks.startWeekRangeOfYear(year);


                log.trace("year=[{}], startWeek=[{}]", year, startWeekRange.startDayStart());

                assertThat(new Duration(Times.asDate(year - 1, 12, 28), startWeekRange.startDayStart()).getStandardDays()).isGreaterThan(0);
                assertThat(new Duration(Times.asDate(year, 1, 3), startWeekRange.endDayStart()).getStandardDays()).isGreaterThan(0);
            }
        });
    }

    @Test
    public void getEndYearAndWeekTest() {
        JParallels.run(1980, 2200, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {
                YearWeek yw = Weeks.endYearAndWeek(year);

                assertThat(year).isEqualTo(yw.weekyear());
                assertThat(yw.weekOfWeekyear()).isGreaterThanOrEqualTo(52);
            }
        });
    }

    @Test
    public void getEndWeekRangeOfYear() {
        JParallels.run(2000, 2100, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {
                WeekRange startWeekRange = Weeks.startWeekRangeOfYear(year);
                WeekRange endWeekRange = Weeks.endWeekRangeOfYear(year - 1);


                log.trace("year=[{}], startWeek=[{}], endWeek=[{}]",
                          year, startWeekRange.startDayStart(), endWeekRange.startDayStart());

                assertThat(new Duration(Times.asDate(year - 1, 12, 28), startWeekRange.startDayStart()).getStandardDays()).isGreaterThan(0);
                assertThat(new Duration(Times.asDate(year, 1, 3), startWeekRange.endDayStart()).getStandardDays()).isGreaterThan(0);

                assertThat(endWeekRange.startDayStart().plusWeeks(1)).isEqualTo(startWeekRange.startDayStart());
                assertThat(endWeekRange.endDayStart().plusDays(1)).isEqualTo(startWeekRange.startDayStart());

            }
        });
    }

    @Test
    public void getWeekRangeTest() {
        JParallels.run(2000, 2100, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {
                DateTime endDay = Times.endTimeOfYear(year - 1);
                DateTime startDay = Times.startTimeOfYear(year);

                YearWeek endDayYearWeek = Weeks.yearAndWeek(endDay);
                assertThat(endDayYearWeek.weekyear()).isGreaterThanOrEqualTo(year - 1);

                YearWeek startDayYearWeek = Weeks.yearAndWeek(startDay);
                assertThat(startDayYearWeek.weekyear()).isLessThanOrEqualTo(year);

                // 해당일자가 속한 주차의 일자들을 구한다. 년말/년초 구간은 꼭 7일이 아닐 수 있다.
                WeekRange endDayWeekRange = Weeks.weekRange(endDayYearWeek);
                WeekRange startDayWeekRange = Weeks.weekRange(startDayYearWeek);

                assertThat(endDayWeekRange.hasPeriod()).isTrue();
                assertThat(startDayWeekRange.hasPeriod()).isTrue();

                log.trace("start day weeksView=[{}]", startDayWeekRange);

                if (endDayYearWeek.equals(startDayYearWeek)) {
                    assertThat(startDayWeekRange).isEqualTo(endDayWeekRange);
                } else {
                    assertThat(startDayWeekRange).isNotEqualTo(endDayWeekRange);
                }
            }
        });
    }

    @Test
    public void addWeekOfYearsTest() {
        JParallels.run(2000, 2100, new JAction1<Integer>() {
            @Override
            public void perform(Integer year) {

                final int step = 2;
                final int maxAddWeeks = 40;

                YearWeek prevResult = null;
                YearWeek maxWeek = Weeks.endYearAndWeek(year);

                for (int week = 1; week < maxWeek.weekOfWeekyear(); week += step) {
                    for (int addWeeks = -maxAddWeeks; addWeeks <= maxAddWeeks; addWeeks += step) {
                        YearWeek current = new YearWeek(year, week);
                        YearWeek result = Weeks.addWeekOfYears(current, addWeeks);

                        if (addWeeks != 0 && prevResult != null) {
                            if (result.weekyear() == prevResult.weekyear())
                                assertThat(result.weekOfWeekyear()).isEqualTo(prevResult.weekOfWeekyear() + step);
                        }

                        assertThat(result.weekOfWeekyear()).isGreaterThan(0);
                        assertThat(result.weekOfWeekyear()).isLessThanOrEqualTo(TimeSpec.MaxWeeksPerYear);

                        prevResult = result;
                    }
                }
            }
        });
    }
}
